### PR TITLE
Create `Finetuner`, `Experiment` and `Run` classes

### PR DESCRIPTION
**Pull request by jupyterjazz**
Closes https://github.com/jina-ai/finetuner.fit/issues/215, https://github.com/jina-ai/finetuner.fit/issues/190 and https://github.com/jina-ai/finetuner.fit/issues/191

After domain names are ready and the "Invalid authentication credentials" error is fixed on the API-side, we can start writing tests and thinking about a proper error handling on the client-side.

### CODE REVIEW

File: .github/workflows/tag.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ name: Release CD
 on:
   push:
     tags:
-      - "v*" # push to version tags trigger the build
+      - 'v*' # push to version tags trigger the build
```
Comment (gmastrapas) [inline .github/workflows/tag.yml:6]: Can you revert these changes here? Replace single quote with double quote? Just to be safe. I'm talking about this file not the py files

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,14 +1,14 @@
 repos:
   - repo: https://github.com/ambv/black
-    rev: 20.8b1
+    rev: '22.3.0'
```
Comment (gmastrapas) [inline .pre-commit-config.yaml:3]: Same for this file

File: codecov.yml [hunk 3]
```diff
@@ -8,8 +8,8 @@ coverage:
         informational: true
         target: auto  # auto compares coverage to the previous base commit
   comment:
-    layout: "reach, diff, flags, files"
+    layout: 'reach, diff, flags, files'
```
Comment (gmastrapas) [inline codecov.yml:11]: same here

File: finetuner/__init__.py [hunk 4]
```diff
@@ -1,9 +1,72 @@
...
+    batch_size: int = 8,
+    freeze: bool = False,
+    output_dim: Optional[int] = None,
+    multi_modal: bool = False,
+    image_modality: Optional[str] = None,
+    text_modality: Optional[str] = None,
+    cpu: bool = False,
+    wandb_api_key: Optional[str] = None,
+):
+    """Start a finetuner run!"""
```
Comment (gmastrapas) [inline finetuner/__init__.py:50]: Can you document the arguments here?

**Review by gmastrapas — COMMENTED** [review]
Some initial comments

File: finetuner/__init__.py [hunk 5]
```diff
@@ -1,9 +1,72 @@
...
 
-load_dotenv()
+def list_models():
+    """List available models for training."""
+    return [
+        'resnet50',
+        'resnet152',
+        'efficientnet_b0',
+        'efficientnet_b4',
+        'bert-base-uncased',
```
[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/__init__.py]: Yeah, I just used this list - https://www.notion.so/User-Journey-for-Finetuner-Client-API-5a5adc556b734e57ba1240f93bdc7dc9#1c9e2e628b714fa0a0a12726846d122d
Updated.

File: finetuner/experiment.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    ) -> Dict[str, Any]:
+        """Create config for a run.
+
+        Note: not a final version yet.
+        :param model: Name of the model to be fine-tuned.
+        :param train_data: Either a `DocumentArray` for training data or a
+                           name of the `DocumentArray` that is pushed on Hubble.
+        :return: Run parameters wrapped up as a config dict.
+        """
+        config = {}
```
Comment (gmastrapas) [inline finetuner/experiment.py:138]: you can write this as a literal:
```
return {
    MODEL: model,
    DATA: {
        ...
    }
}
```

**Review by gmastrapas — COMMENTED** [review]
Round 2

File: finetuner/client/client.py [hunk 7]
```diff
@@ -1,33 +1,31 @@
...
 
-class Client(BaseClient):
+class FinetunerV1Client(_BaseClient):
+    """
+    The Finetuner v1 API client.
+    """
+
+    """ Experiment API """
+
     def create_experiment(self, name: str, **kwargs) -> dict:
```
[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/client/client.py]: That's supposed to be for `description`. Changed it.

File: finetuner/run.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        raise NotImplementedError('Not yet implemented, stay tuned!')
+
+    def save_model(self, path: str = FINETUNED_MODELS_DIR):
+        """Save model(s) if the run is finished.
+
+        :param path: Directory where the model(s) will be stored.
+        :returns: A list of str object(s) that indicate the download path.
+        """
+        if self.status() == FINISHED:
+            download_path = self._client.download_model(
```
[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/run.py:57]: We haven't tested pulling the models yet, but looks correct to me

[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/run.py:57]: Oops. changed this too.

File: finetuner/__init__.py [hunk 9]
```diff
@@ -1,9 +1,95 @@
-import hubble
+from typing import Optional, Union
+
+from docarray import DocumentArray
 from dotenv import load_dotenv
 
+from finetuner.finetuner import Finetuner
+
+load_dotenv()
```
[the remark this answers is not in the document]
Comment (gmastrapas) [reply finetuner/__init__.py:8]: can we just parse `os.environ`? We have an extra requirement for no reason

[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/__init__.py:8]: I believe first we need to load these environment variables, that's why we use `load_dotenv()`

[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/__init__.py:8]: We don't need it here because we're using github secrets.. How do you suggest loading envs then?

[the remark this answers is not in the document]
Comment (gmastrapas) [reply finetuner/__init__.py:8]: By loading you mean setting the env vars right?

File: finetuner/__init__.py [hunk 10]
```diff
@@ -1,9 +1,95 @@
...
+    :param experiment_name: Name of the experiment.
+    :param loss: Name of the loss function used for fine-tuning.
+    :param miner: Name of the miner to create tuple indices for the loss function.
+    :param optimizer: Name of the optimizer used for fine-tuning.
+    :param learning_rate: learning rate for the optimizer.
+    :param epochs: Number of epochs for fine-tuning.
+    :param batch_size: Number of items to include in a batch.
+    :param freeze: If set to True, will freeze all layers except the last one.
+    :param output_dim: The expected output dimension.
+                       If set, will attach a projection head.
```
Comment (gmastrapas) [inline finetuner/__init__.py:66]: single tab

File: finetuner/__init__.py [hunk 11]
```diff
@@ -1,9 +1,95 @@
...
+    :param multi_modal: Boolean value, if `True`,
+                        required argument to the `DataLoader`.
```
Comment (gmastrapas) [inline finetuner/__init__.py:68]: single tab

File: finetuner/hubble.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    :param train_data: Either a `DocumentArray` for training data that needs to be
+        pushed on Hubble or a name of the `DocumentArray` that is already uploaded.
+    :param eval_data: Either a `DocumentArray` for evaluation data that needs to be
+        pushed on Hubble or a name of the `DocumentArray` that is already uploaded.
+    :param experiment_name: Name of the experiment.
+    :param run_name: Name of the run.
+    :returns: Name(s) of pushed `DocumentArray`-s.
+    """
+    if isinstance(train_data, DocumentArray):
+        da_name = '-'.join(
```
Comment (gmastrapas) [inline finetuner/hubble.py:30]: how about this function is only about a single document array and then you call it twice?

File: finetuner/hubble.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
+from typing import List, Optional, Tuple, Union
+
+from docarray import DocumentArray
+
+from finetuner.client import FinetunerV1Client
+from finetuner.constants import EVAL_DATA, FINETUNED_MODELS_DIR, MODEL_IDS, TRAIN_DATA
+
+
+def push_data_to_hubble(
+    client: FinetunerV1Client,
```
Comment (gmastrapas) [inline finetuner/hubble.py:10]: we pass this client everywhere. How about we initialize it once?

File: finetuner/run.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            download_path = download_model(
+                client=self._client,
+                experiment_name=self.experiment_name,
+                run_name=self.name,
+                path=path,
+            )
+        else:
+            # tell users that the run hasn't finished yet, thus they can't download
+            # the model.
+            pass
```
Comment (gmastrapas) [inline finetuner/run.py:67]: you can raise an exception here

File: finetuner/run.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        return download_path
```
Comment (gmastrapas) [inline finetuner/run.py:68]: if status is not finished, the referenced variable does not exist

File: finetuner/run.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    """
+
+    def __init__(
+        self,
+        client: FinetunerV1Client,
+        name: str,
+        experiment_name: str,
+        config: dict,
+        created_at: str,
+        description: Optional[str] = '',
```
Comment (gmastrapas) [inline finetuner/run.py:26]: `str`

File: finetuner/experiment.py [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        return self._name
+
+    def get_run(self, name: str) -> Run:
+        """Get a run by its name.
+
+        :param name: Name of the run.
+        :return: A `Run` object.
+        """
+        run_info = self._client.get_run(experiment_name=self._name, run_name=name)
+        run = Run(name=run_info[NAME], experiment_name=self._name, client=self._client)
```
Comment (gmastrapas) [inline finetuner/experiment.py:56]: this will fail here, you are omitting required arguments

File: finetuner/experiment.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        return run
+
+    def list_runs(self) -> List[Run]:
+        """List every run inside the experiment.
+
+        :return: List of `Run` objects.
+        """
+        run_infos = self._client.list_runs(experiment_name=self._name)
+        runs = [
+            Run(name=run_info[NAME], experiment_name=self._name, client=self._client)
```
Comment (gmastrapas) [inline finetuner/experiment.py:66]: this will fail here, you are omitting required arguments

File: finetuner/finetuner.py [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+        return experiment
+
+    def create_experiment(self, name: Optional[str] = None) -> Experiment:
+        """Create an experiment.
+
+        :param: Optional name of the experiment.
+        :returns: An `Experiment` object.
+        """
+        if not name:
+            name = get_random_name()
```
Comment (gmastrapas) [inline finetuner/finetuner.py:46]: you need to use the working directory as the experiment name

File: finetuner/finetuner.py [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+        return experiment
+
+    def get_experiment(self, name: str) -> Experiment:
+        """Get an experiment by its name.
+
+        :param name: Name of the experiment.
+        :returns: An `Experiment` object.
+        """
+        experiment_info = self._client.get_experiment(name=name)
+        experiment = Experiment(name=experiment_info[NAME], client=self._client)
```
Comment (gmastrapas) [inline finetuner/finetuner.py:64]: this will fail, you are omitting required arguments

File: finetuner/finetuner.py [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+        return experiment
+
+    def list_experiments(self) -> List[Experiment]:
+        """List every experiment."""
+        experiment_infos = self._client.list_experiments()
+        experiments = [
```
Comment (gmastrapas) [inline finetuner/finetuner.py:70]: instead of creating a variable just to return it, you can do it in-place:
```
return [
            Experiment(
                client=self._client,
                name=experiment_info[NAME],
                status=experiment_info[STATUS],
                created_at=experiment_info[CREATED_AT],
            )
            for experiment_info in experiment_infos
        ]
```
I think it looks cleaner

**Review by gmastrapas — CHANGES_REQUESTED** [review]

File: finetuner/experiment.py (hunk 17 above)
Comment (jupyterjazz) [reply finetuner/experiment.py:56]: yes, I found these when I started testing. I'll change it in the next PR.

File: finetuner/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jupyterjazz) [reply finetuner/__init__.py:8]: No, I got what you mean. We also have `path` in the requirements which is definitely not needed. How about leaving them for now and refactoring in a few days?
